Add `@inherit` service to get inherited behavior values (#1887)
plone/plone.restapi: docs/source/endpoints/inherit.md

---
myst:
  html_meta:
    "description": "The inherit endpoint shows values inherited from content items higher in the hierarchy."
    "property=og:description": "The inherit endpoint shows values inherited from content items higher in the hierarchy."
    "property=og:title": "Inherit behaviors"
    "keywords": "Plone, plone.restapi, REST, API, inherit, acquisition, behaviors"
---

(inherit-behaviors-label)=

# Inherit behaviors

> The `@inherit` service makes it possible to inherit data from behaviors defined on ancestors in the object hierarchy.

Author: @stevepiercy @stevepiercy What about something like this?

Plone content is arranged in a hierarchy: a content item has a parent, which has its own parent, all the way up to the Plone site root.
Together, all these parents are _ancestors_.
The `@inherit` service makes it possible to access data from a behavior defined on one of these ancestors.

> The `@inherit` service makes it possible to inherit data from behaviors defined on ancestors in the object hierarchy.

Author: Done in https://github.com/plone/plone.restapi/pull/1887/commits/c6581193b0d45bc5d77e8f44b4904bdbb34e8f27

> The `@inherit` service makes it possible to inherit data from behaviors defined on ancestors in the object hierarchy.

Reviewer: > I disagree. I think the verbosity makes it harder to understand for non-native English readers.

This is a wrong assumption, as long simple and different words are used. Using different words for the same makes it easier to get context and understand. Just @davisagli put it in a series of really simple sentences, with is great and improves it a lot.

```{tip}
Inheriting behaviors is similar to the concept of {term}`acquisition` in Zope, but it doesn't happen automatically, so it's safer.
```

To use the service, send a `GET` request to the `@inherit` endpoint in the context of the content item that is the starting point for inheriting.
Specify the `expand.inherit.behaviors` parameter as a comma-separated list of behaviors.

```{eval-rst}
..  http:example:: curl httpie python-requests
    :request: ../../../src/plone/restapi/tests/http-examples/inherit_get.req
```

For each behavior, the service will find the closest ancestor which provides that behavior.

> The result includes `from` (the `@id` and `title` of the item from which values were inherited)
> and `data` (values for any fields that are part of the behavior).

Reviewer 2: ```suggestion
The result includes `from` (the `@id` and `title` of the item from which values were inherited) and `data` (values for any fields that are part of the behavior).
```

```{literalinclude} ../../../src/plone/restapi/tests/http-examples/inherit_get.resp
:language: http
```

> Ancestor items for which the current user lacks the View permission will be skipped.

Reviewer 2: ```suggestion
Ancestor items for which the current user lacks the view permission will be skipped.
```

(inherit-behaviors-expansion-label)=

## Expansion

This endpoint can be used with the {doc}`../usage/expansion` mechanism which allows getting more information about a content item in one query, avoiding unnecessary requests.

> You can make a `GET` request for a content item, and include parameters to request `inherit` expansion for specific behaviors:

Reviewer 2: ```suggestion
You can make a `GET` request for a content item, and include parameters to request `inherit` expansion for specific behaviors.
```

Reviewer 2: Not worth the effort. Documentation follows https://6.docs.plone.org/contributing/documentation/authors.html#vale-for-american-english-spelling-grammar-and-syntax-and-style-guide, specifically https://learn.microsoft.com/en-us/style-guide/punctuation/colons. Maybe next time.

Author: @stevepiercy I can't find a specific reference to this in the Microsoft style guide, but Microsoft's Azure REST API docs also use a colon to introduce request/response examples, for example in https://learn.microsoft.com/en-us/rest/api/azure/#send-the-request

Reviewer 2: Sometimes docs don't follow their own style guide, like that _never_ happens in Plone. 😏

```{eval-rst}
..  http:example:: curl httpie python-requests
    :request: ../../../src/plone/restapi/tests/http-examples/inherit_expansion.req
```

> The response will include data from the `@inherit` endpoint within the `@components` property:

Reviewer 2: ```suggestion
The response will include data from the `@inherit` endpoint within the `@components` property.
```

---
myst:
  html_meta:
    "description": "The inherit endpoint shows values inherited from content items higher in the hierarchy."
    "property=og:description": "The inherit endpoint shows values inherited from content items higher in the hierarchy."
    "property=og:title": "Inherit behaviors"
    "keywords": "Plone, plone.restapi, REST, API, inherit, acquisition, behaviors"
---

(inherit-behaviors-label)=

# Inherit behaviors

> Plone content is arranged in a hierarchy: a content item has a parent, which has its own parent, all the way up to the Plone site root.
> Together, all these parents are _ancestors_.

Reviewer 2: ```suggestion
Plone content items are arranged in a hierarchy.
Each content item has a parent, each of which may have its own parent, and continue all the way to the Plone site root.
Together, the chain of parents are _ancestors_ of the content item.
```

Author: @stevepiercy Sorry, I'm in a bit of a hurry today. Your edits are good. Committed directly to `main` in https://github.com/plone/plone.restapi/commit/8816d589c3fae1fbbb48378a345b179e4f38e15c
